libeos-update-server/repo.c

```diff
@@ -680,18 +681,33 @@ serve_file_if_exists (SoupMessage *msg,
     }
 
   mapping = g_mapped_file_new (raw_path, FALSE, &error);
-  if (mapping == NULL)
+  if (mapping != NULL)
+    file_bytes = g_mapped_file_get_bytes (mapping);
+  else
     {
-      g_warning ("Failed to map %s: %s", raw_path, error->message);
-      soup_message_set_status (msg, SOUP_STATUS_INTERNAL_SERVER_ERROR);
-      return FALSE;
+      g_autofree guint8 *contents = NULL;
+      gsize contents_len = 0;
+
+      /* mmap() can legitimately fail if the underlying file system doesn’t
+       * support it, which can happen if we’re using an overlayfs. Fall back to
+       * reading in the file. */
+      g_clear_error (&error);
+      if (!g_file_get_contents (raw_path, (gchar **) &contents, &contents_len, &error))
+        {
+          g_warning ("Failed to load ‘%s’: %s", raw_path, error->message);
+          soup_message_set_status (msg, SOUP_STATUS_INTERNAL_SERVER_ERROR);
+          return FALSE;
```

Comment:
We should have `*served = FALSE;` here, just to be safe even though all the callers initialize `served`.

Reply:
I don’t think so — the code didn’t used to do this, and the convention in GLib is for all out-parameters to remain unchanged if the function fails. Since we can guarantee all the callers behave properly here (thanks for checking), I don’t want to include extra code to deal with the non-existent programmer error of if they don’t.

Reply:
Oh, I wasn't aware of that convention